hazelcast/src/main/java/com/hazelcast/spi/Offload.java

```diff
@@ -0,0 +1,142 @@
+     *
+     * @return the Operation.
+     */
+    public final Operation source() {
+        return source;
+    }
+
+    /**
+     * Initializes the Offload.
+     *
+     * As part of the initialization, the {@link OperationResponseHandler} of the source {@link Operation} is replaced by
+     * a decorated version that takes care of automatic deregistration of the operation on completion.
+     *
+     * This method is called before the {@link #start()} is called by the Operation infrastructure. An implementor of the
+     * {@link Offload} doesn't need to deal with this method.
+     *
+     * @param nodeEngine the {@link NodeEngineImpl}
+     */
+    public final void init(NodeEngineImpl nodeEngine, Set<Operation> asyncOperations) {
+        this.nodeEngine = nodeEngine;
+        this.operationService = nodeEngine.getOperationService();
+        this.serializationService = nodeEngine.getSerializationService();
+        this.asyncOperations = asyncOperations;
+        this.executionService = nodeEngine.getExecutionService();
+
+        asyncOperations.add(source);
+
+        if (shouldDecorateOperationResponseHandler()) {
+            // todo: it could happen that when an operation is offload multiple times, that you get a chain of offload
+            // response handlers.
+            source.setOperationResponseHandler(new OperationResponseHandlerImpl());
+        }
+    }
+
+    private boolean shouldDecorateOperationResponseHandler() {
+        OperationResponseHandler currentHandler = source.getOperationResponseHandler();
+        if (currentHandler instanceof OperationResponseHandlerImpl) {
+            OperationResponseHandlerImpl o = (OperationResponseHandlerImpl) currentHandler;
+
+            // todo: if true is returned, you can still get a chain of different offload
+            return o.offload != this;
+        } else {
+            return true;
+        }
+    }
+
+    /**
+     * Starts the actual offloading.
+     *
+     * This method is still called on the same thread that called the {@link Operation#call()} where this Offload
+     * instance was returned. So in most cases you want to schedule some work and then return from this method ASAP so
+     * that the thread is released.
+     *
+     * This method is called after the {@link Operation#afterRun()} is called.
+     *
+     * It is allowed to call {@link Operation#sendResponse(Object)} in the start method if there is nothing to offload.
+     *
+     * @throws Exception if something fails. If this happens, regular Operation exception handling is triggered and normally
+     *                   the exception is returned to the caller.
+     */
+    public abstract void start() throws Exception;
+
+    private class OperationResponseHandlerImpl implements OperationResponseHandler {
+        private final Offload offload = Offload.this;
+
+        @Override
+        public void sendResponse(Operation op, Object response) {
```

Comment:
Will this always be called?

Reply:
As long as the user has not changed the response handler of the 'source' operation.

Reply:
I have improved the documentation. It explicitly states that in case of an Operation that is offloaded, the response handler of the source operation should not be modified.

I have not seen any violations of this. The response handlers of the forked operations are intercepted; but the response handler of the 'source' operation remains is not changed.